src/main/clojure/clara/rules/compiler.clj

```diff
@@ -1242,19 +1248,196 @@
+                                                        ;; See PR 243 for more detailed discussion
+                                                        (select-keys cmeta [:line :file]))
+                                             id
+                                             :alpha-expr
+                                             {:file (or (:file cmeta) *file*)
+                                              :compile-ctx {:condition condition
+                                                            :env env
+                                                            :msg "compiling alpha node"}})))
+                            {}
+                            alpha-graph)
+        id->expr (reduce-kv (fn [prev id production-node]
+                             (let [production (-> production-node :production)]
+                               (handle-expr prev
+                                            (with-meta (compile-action (:bindings production-node)
+                                                                       (:rhs production)
+                                                                       (:env production))
+                                                       (meta (:rhs production)))
+                                            id
+                                            :action-expr
+                                            ;; ProductionNode expressions can be arbitrary code, therefore we need the
+                                            ;; ns where the production was define so that we can compile the expression
+                                            ;; later.
+                                            {:ns (:ns-name production)
+                                             :file (-> production :rhs meta :file)
+                                             :compile-ctx {:production production
+                                                           :msg "compiling production node"}})))
+                            id->expr
+                           (:id-to-production-node beta-graph))]
+    (reduce-kv
+      (fn [prev id beta-node]
+        (let [condition (:condition beta-node)
+              condition (if (symbol? condition)
+                          (.loadClass (clojure.lang.RT/makeClassLoader) (name condition))
+                          condition)]
+          (case (or (:node-type beta-node)
+                    ;; If there is no :node-type then the node is the ::root-condition
+                    ;; however, creating a case for nil could potentially cause weird effects if something about the
+                    ;; compilation of the beta graph changes. Therefore making an explicit case for ::root-condition
+                    ;; and if there was anything that wasn't ::root-condition this case statement will fail rather than
+                    ;; failing somewhere else.
```

Comment:
I like failing instead of waiting to later and being broken. However, I think a it'd be nicer to have that default `case` at the end that threw a more meaningful exception than "no branch of `case` matched".

eg
```clj
(case x
 y :foo
 z :bar
 (throw (ex-info "explanation"))
```

However, this isn't a critical change for the sake of this review. It's more of a thought about trying to give more meaningful compilation failure errors in the future.